### DISCUSSION TITLE
Convert object to query?

### DISCUSSION THREAD

**Discussion by SebastianStehle** [graphql-dotnet/graphql-dotnet#3677] [Q&A]
Hi,

I have this super ugly code in my tests:
https://github.com/Squidex/squidex/blob/master/backend/tests/Squidex.Domain.Apps.Entities.Tests/Contents/GraphQL/GraphQLTestBase.cs#L167

It converts an object to the query syntax, e.g. for something like that:

```
 var query = CreateQuery(@"
                mutation {
                  createMySchemaContent(data: <DATA>, id: '123', publish: true) {
                    <FIELDS_CONTENT>
                  }
                }", contentId, content);
```

Is there something builtin, that I have not seen yet?

**Comment by Shane32** [comment]
Well, I'm sorta wondering why you are not using variables.  I don't know that it helps, but this is how I set up my tests, using the Shouldly nuget package as a basis:

```cs
    [Fact]
    public async Task Add()
    {
        //seed the database as needed:
        //await Db.SeedAsync<Setting>();

        var ret = await RunQueryAsync(
            """
            mutation ($value: SettingInput!) {
                setting {
                    add (value: $value) {
                        id
                        name
                        value
                    }
                }
            }
            """,
            new {
                value = new {
                    name = "Testing",
                    value = "TestValue",
                }
            });
        // validate that the query did not return any errors
        ret.ShouldBeSuccessful();

        // create an anonymous object containing all the values to validate
        var row = await Db.Settings.Where(x => x.Id == 1).SingleAsync();
        var actual = new { ret, row };
        // convert to json and compare to an approval file stored alongside the tests
        actual.ShouldMatchApproved();
    }
```

But you'll need to set up the test infrastructure yourself.  For my implementation, the biggest benefits are that tests are extremely easy to write and everything is run on a real MS-SQL implementation.  But the biggest problem is the fact that most of the tests rely on consistent ids being generated by the database engine, so the database is regenerated for each test, meaning it's slow, as each test takes a couple hundred milliseconds to run.

If you just want an 'official' way to convert arbitrary data to a graphql literal, I'd probably write code like this:
1. Convert the data to a JSON
2. Convert the JSON back to a set of nested dictionaries (the `Input` type in graphql, which already includes a converter)
3. Convert the object to AST via ComplexScalarGraphType (see PR #3257 )
4. Print the AST

**Reply by SebastianStehle** [reply]
The point with the variables is a very good question  :D

I also have api tests where I test the full flow, btw....

**Comment by gao-artur** [comment]
We use [GraphQlClientGenerator](https://github.com/Husqvik/GraphQlClientGenerator) to generate strongly typed client from our schema. The benefit of this is Intellisense and compile time errors on schema changes.
For the DB dependency, we start a docker container using our own library, but there is also an open source [testcontainers-dotnet](https://github.com/testcontainers/testcontainers-dotnet). For the fast DB cleanup between tests, we use [Respawn](https://github.com/jbogard/Respawn), which can be configured with `new RespawnerOptions { WithReseed = true }` to get stable ids on every test.
